securedrop/tests/functional/functional_test.py

```diff
@@ -33,19 +34,50 @@ class FunctionalTest():
 
     def _unused_port(self):
         s = socket.socket()
-        s.bind(("localhost", 0))
+        s.bind(('127.0.0.1', 0))
         port = s.getsockname()[1]
         s.close()
         return port
 
+
+    def _create_tor_browser_webdriver(self, abs_log_file_path):
+        # Read the TorBrowser location from Ansible
+        with open('/opt/.tbb_path_file') as f:
+           path_to_tbb = f.readline().strip()
+        path_to_tbb = path_to_tbb + os.path.sep + "tor-browser_en-US"
+        path_to_tbb = os.path.abspath(path_to_tbb)
+
+        # Don't use Tor when reading from localhost, and turn off private browsing.
+        # We need to turn off private browsing because we won't be able to access the browser's cookies in private browsing mode.
+        # Since we use session cookies in SD anyway (in private browsing mode all cookies are set as session cookies),
+        # this should not affect session lifetime.
+        pref_dict = {
+                     'network.proxy.no_proxies_on': '127.0.0.1',
+                     'browser.privatebrowsing.autostart': False
+                    }
+        driver = TorBrowserDriver(path_to_tbb,
+                                  pref_dict=pref_dict,
+                                  tbb_logfile_path=abs_log_file_path)
+        return driver
+
+    def _create_firefox_webdriver(self, abs_log_file_path):
+        with open(abs_log_file_path, 'a') as f:
+            firefox = firefox_binary.FirefoxBinary(log_file=f)
+            return webdriver.Firefox(firefox_binary=firefox)
+
     def _create_webdriver(self):
-        log_file = open(join(LOG_DIR, 'firefox.log'), 'a')
-        log_file.write(
-            '\n\n[%s] Running Functional Tests\n' % str(
-                datetime.now()))
-        log_file.flush()
-        firefox = firefox_binary.FirefoxBinary(log_file=log_file)
-        return webdriver.Firefox(firefox_binary=firefox)
+        abs_log_file_path = os.path.abspath(join(dirname(__file__), '../log/firefox.log'))
+        with open(abs_log_file_path, 'a') as f:
+            log_msg = '\n\n[%s] Running Functional Tests\n' % str(datetime.now())
+            f.write(log_msg)
+            f.flush()
```

Comment:
You shouldn't need flush here. Python will automatically flush here before closing the file.